plugins/coverimage.koplugin/main.lua

```diff
@@ -0,0 +1,190 @@
+local Device = require("device")
+
+if not Device.isAndroid() and not Device.isEmulator() then
+    return { disabled = true }
+end
+
+local DocSettings = require("docsettings")
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local logger = require("logger")
+local _ = require("gettext")
+local T = require("ffi/util").template
+
+local CoverImage = WidgetContainer:new{
+    name = 'coverimage',
+    is_doc_only = true,
+    settings_id = 0,
+}
+
+function CoverImage:_enabled()
+    return self.enabled
+end
+
+function CoverImage:_restore()
+    return self.enabled and self.restore
+end
+
+-- check if settings of lastfile prohibits creation of cover image
+function CoverImage:excluded()
+    local lastfile = G_reader_settings:readSetting("lastfile")
+    local exclude_ss = false -- consider it not excluded if there's no docsetting
+    if DocSettings:hasSidecarFile(lastfile) then
+        if self and self.ui then
```

Comment:
No need to check `if self` (it always exists), I guess you meant `if self.ui and self.ui.doc_settings then`.

But I still don't like all that :)
It feels none of this is needed, and you don't need to care about `lastfile`:
- in `CoverImage:onCloseDocument()`, you don't really need to know if `excluded`: if `restore` and there's a .bak: restore it.
- in `CoverImage:onReaderReady(doc_settings)`, you have the doc_settings, so just read it

If there's some edge case because you call these 2 methods when toggling stuff in the menu, may be you should have 2 other methods: `CoverImage:backupPreviousAndCreateNewOne()` and `CoverImage:restorePreviousIfAny()` (with better names :) to do the low level check and action - and onCloseDocument/OnReaderReady/your menu callbacks would do the check they can from what they have - and call the 2 lowlevel methods if stuff should be done.